### ISSUE TITLE
[BUG] Latest available pygit2 from Debian 11 and Pypi break GitFS

### ISSUE THREAD

**Issue by DaAwesomeP** [saltstack/salt#61790]
**Description**
The version of pygit2 included with Debian 11 (v1.4.0) only allows for RSA PEM keys (see [this issue](https://github.com/libgit2/pygit2/issues/1013)) and sends them with SHA-1 regardless of if they are SHA-256 or SHA-512 keys. GitHub has disabled all use of RSA SHA-1 keys.

As of today (3/15/2022), RSA SHA-1 keys are not allowed: https://github.blog/2021-09-01-improving-git-protocol-security-github/

**Setup**
(Please provide relevant configs and/or SLS files (be sure to remove sensitive info. There is no general set-up of Salt.)

Please be as specific as possible and give set-up details.

- [ ] on-prem machine
- [x] VM (Virtualbox, KVM, etc. please specify): VMWare
- [ ] VM running on a cloud service, please be explicit and add details
- [ ] container (Kubernetes, Docker, containerd, etc. please specify)
- [ ] or a combination, please be explicit
- [ ] jails if it is FreeBSD

  
**Steps to Reproduce the behavior**
1. First try a normal RSA key: `ssh-keygen -m PEM -t rsa -b 4096`
2. Also try a SHA-512 RSA key:  `ssh-keygen -m PEM -t rsa -E sha512 -b 4096`
3. Also try a SHA-256 RSA key:  `ssh-keygen -m PEM -t rsa -E sha256 -b 4096`
4. Also try an ed25519 key:  `ssh-keygen -m PEM -a 100 -t ed25519 -E sha512`
5. Configure these keys as deploy keys for a private repo on GitHub
6. Configure a remote repo:
   ```yaml
   fileserver_backend:
     - gitfs
   
   gitfs_provider: pygit2
   gitfs_remotes:
     - ssh://git@github.com/myorg/my-private-repo.git:
       - base: main
       - pubkey: /root/.ssh/id_my-key.pub
       - privkey: /root/.ssh/id_my-key
   ```
7. Restart salt master
8. Try to pull `salt-run fileserver.update`

With any RSA key, you will see:
```
[ERROR   ] Error occurred fetching gitfs remote 'ssh://git@github.com/myorg/my-private-repo.git': ERROR: You're using an RSA key with SHA-1, which is no longer allowed. Please use a newer client or a different key type.
Please see https://github.blog/2021-09-01-improving-git-protocol-security-github/ for more information.

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/salt/utils/gitfs.py", line 1870, in _fetch
    fetch_results = origin.fetch(**fetch_kwargs)
  File "/usr/lib/python3/dist-packages/pygit2/remote.py", line 128, in fetch
    payload.check_error(err)
  File "/usr/lib/python3/dist-packages/pygit2/callbacks.py", line 93, in check_error
    check_error(error_code)
  File "/usr/lib/python3/dist-packages/pygit2/errors.py", line 65, in check_error
    raise GitError(message)
_pygit2.GitError: ERROR: You're using an RSA key with SHA-1, which is no longer allowed. Please use a newer client or a different key type.
Please see https://github.blog/2021-09-01-improving-git-protocol-security-github/ for more information.
```

With any non-RSA key you will see:
```
[ERROR   ] Error occurred fetching gitfs remote 'ssh://git@github.com/myorg/my-private-repo.git': callback failed to initialize SSH credentials
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/salt/utils/gitfs.py", line 1870, in _fetch
    fetch_results = origin.fetch(**fetch_kwargs)
  File "/usr/lib/python3/dist-packages/pygit2/remote.py", line 128, in fetch
    payload.check_error(err)
  File "/usr/lib/python3/dist-packages/pygit2/callbacks.py", line 93, in check_error
    check_error(error_code)
  File "/usr/lib/python3/dist-packages/pygit2/errors.py", line 65, in check_error
    raise GitError(message)
_pygit2.GitError: callback failed to initialize SSH credentials
```

**Expected behavior**

**Screenshots**

**Versions Report**
<details><summary>salt --versions-report</summary>
(Provided by running salt --versions-report. Please also mention any differences in master/minion versions.) 

```yaml
Salt Version:
          Salt: 3004
 
Dependency Versions:
          cffi: Not Installed
      cherrypy: Not Installed
      dateutil: 2.8.1
     docker-py: Not Installed
         gitdb: 4.0.5
     gitpython: 3.1.14
        Jinja2: 2.11.3
       libgit2: 1.1.0
      M2Crypto: Not Installed
          Mako: Not Installed
       msgpack: 1.0.0
  msgpack-pure: Not Installed
  mysql-python: Not Installed
     pycparser: Not Installed
      pycrypto: Not Installed
  pycryptodome: 3.9.7
        pygit2: 1.4.0
        Python: 3.9.2 (default, Feb 28 2021, 17:03:44)
  python-gnupg: Not Installed
        PyYAML: 5.3.1
         PyZMQ: 20.0.0
         smmap: 4.0.0
       timelib: Not Installed
       Tornado: 4.5.3
           ZMQ: 4.3.4
 
System Versions:
          dist: debian 11 bullseye
        locale: utf-8
       machine: x86_64
       release: 5.10.0-12-amd64
        system: Linux
       version: Debian GNU/Linux 11 bullseye
```
</details>

**Additional context**

**Comment by DaAwesomeP** [comment]
I have confirmed that running `ssh -i /path/to/key -T git@github.com` works for these key methods on this machine, so the issue is not with the keys.

**Comment by DaAwesomeP** [comment]
This may be related to https://github.com/saltstack/salt/issues/46159

**Comment by DaAwesomeP** [comment]
@whytewolf Doesn't seem to help:
```bash
$ apt install python3-pip
$ apt purge python3-pygit2
$ apt autoremove # removed libgit2-1.1 libhttp-parser2.9 libmbedcrypto3 libmbedtls12 libmbedx509-0 python3-cached-property python3-cffi-backend
$ python3 -m pip install pygit2
$ systemctl restart salt-master
```
```bash
$ salt-run fileserver.update
[ERROR   ] Unable to connect pusher: Stream is closed
[ERROR   ] Error occurred fetching gitfs remote 'ssh://git@github.com/myorg/my-private-repo.git': ERROR: You're using an RSA key with SHA-1, which is no longer allowed. Please use a newer client or a different key type.
Please see https://github.blog/2021-09-01-improving-git-protocol-security-github/ for more information.

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/salt/utils/gitfs.py", line 1870, in _fetch
    fetch_results = origin.fetch(**fetch_kwargs)
  File "/usr/local/lib/python3.9/dist-packages/pygit2/remote.py", line 147, in fetch
    payload.check_error(err)
  File "/usr/local/lib/python3.9/dist-packages/pygit2/callbacks.py", line 93, in check_error
    check_error(error_code)
  File "/usr/local/lib/python3.9/dist-packages/pygit2/errors.py", line 65, in check_error
    raise GitError(message)
_pygit2.GitError: ERROR: You're using an RSA key with SHA-1, which is no longer allowed. Please use a newer client or a different key type.
Please see https://github.blog/2021-09-01-improving-git-protocol-security-github/ for more information.


[ERROR   ] Unable to connect pusher: Stream is closed
```

**Comment by DaAwesomeP** [comment]
I attempted to use GitPython instead, but I cannot since I have separate repos for salt and pillar. GitHub requires deploy keys be unique and separate per repo.

**Comment by whytewolf** [comment]
don't use the rsa key.

**Comment by DaAwesomeP** [comment]
Using pygit2 from pip and an ed25519 key:
```
salt-run fileserver.update
[ERROR   ] Unable to connect pusher: Stream is closed

Exception ignored from cffi callback <function _credentials_cb at 0x7f9c5f71ef70>:
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/dist-packages/pygit2/callbacks.py", line 326, in wrapper
    @wraps(f)
  File "/usr/lib/python3/dist-packages/salt/scripts.py", line 61, in _handle_signals
    sys.exit(salt.defaults.exitcodes.EX_OK)
SystemExit: 0
[ERROR   ] Error occurred fetching gitfs remote 'ssh://git@github.com/myorg/my-private-repo.git': callback failed to initialize SSH credentials
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/salt/utils/gitfs.py", line 1870, in _fetch
    fetch_results = origin.fetch(**fetch_kwargs)
  File "/usr/local/lib/python3.9/dist-packages/pygit2/remote.py", line 147, in fetch
    payload.check_error(err)
  File "/usr/local/lib/python3.9/dist-packages/pygit2/callbacks.py", line 93, in check_error
    check_error(error_code)
  File "/usr/local/lib/python3.9/dist-packages/pygit2/errors.py", line 65, in check_error
    raise GitError(message)
_pygit2.GitError: callback failed to initialize SSH credentials
[ERROR   ] Unable to connect pusher: Stream is closed
True
```

**Comment by whytewolf** [comment]
can you list what version of pygit2 you are using now?

**Comment by whytewolf** [comment]
did the pip install show that it was using the wheel? or did in compile it?

**Comment by DaAwesomeP** [comment]
OK, tried it with force-compiled but the Debian 11 (latest stable) `libgit2-dev` is too old for it:
```
$ apt install libgit2-dev
$ python3 -m pip install --force --no-binary pygit2 pygit2
Collecting pygit2
  Using cached pygit2-1.9.0.tar.gz (290 kB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Installing backend dependencies ... done
    Preparing wheel metadata ... done
Collecting cffi>=1.9.1
  Using cached cffi-1.15.0-cp39-cp39-manylinux_2_12_x86_64.manylinux2010_x86_64.whl (444 kB)
Collecting pycparser
  Using cached pycparser-2.21-py2.py3-none-any.whl (118 kB)
Building wheels for collected packages: pygit2
  Building wheel for pygit2 (PEP 517) ... error
  ERROR: Command errored out with exit status 1:
   command: /usr/bin/python3 /tmp/tmpida1o9el_in_process.py build_wheel /tmp/tmp5wvetfwi                                                                                                                                                                                       
       cwd: /tmp/pip-install-mw_5hm13/pygit2_d0cf820d050b4656a2b04b22fbe310cd                                                                                                                                                                                                  
  Complete output (65 lines):                                                                                                                                                                                                                                                  
  running bdist_wheel                                                                                                                                                                                                                                                          
  running build                                                                                                                                                                                                                                                                
  running build_py                                                                                                                                                                                                                                                             
  creating build                                                                                                                                                                                                                                                               
  creating build/lib.linux-x86_64-3.9                                                                                                                                                                                                                                          
  creating build/lib.linux-x86_64-3.9/pygit2                                                                                                                                                                                                                                   
  copying pygit2/__init__.py -> build/lib.linux-x86_64-3.9/pygit2
  copying pygit2/credentials.py -> build/lib.linux-x86_64-3.9/pygit2
  copying pygit2/submodule.py -> build/lib.linux-x86_64-3.9/pygit2
  copying pygit2/settings.py -> build/lib.linux-x86_64-3.9/pygit2
  copying pygit2/_run.py -> build/lib.linux-x86_64-3.9/pygit2
  copying pygit2/errors.py -> build/lib.linux-x86_64-3.9/pygit2
  copying pygit2/repository.py -> build/lib.linux-x86_64-3.9/pygit2
  copying pygit2/utils.py -> build/lib.linux-x86_64-3.9/pygit2
  copying pygit2/config.py -> build/lib.linux-x86_64-3.9/pygit2
  copying pygit2/refspec.py -> build/lib.linux-x86_64-3.9/pygit2
  copying pygit2/remote.py -> build/lib.linux-x86_64-3.9/pygit2
  copying pygit2/packbuilder.py -> build/lib.linux-x86_64-3.9/pygit2
  copying pygit2/index.py -> build/lib.linux-x86_64-3.9/pygit2
  copying pygit2/callbacks.py -> build/lib.linux-x86_64-3.9/pygit2
  copying pygit2/blame.py -> build/lib.linux-x86_64-3.9/pygit2
  copying pygit2/_build.py -> build/lib.linux-x86_64-3.9/pygit2
  copying pygit2/ffi.py -> build/lib.linux-x86_64-3.9/pygit2
  creating build/lib.linux-x86_64-3.9/pygit2/decl
  copying pygit2/decl/common.h -> build/lib.linux-x86_64-3.9/pygit2/decl
  copying pygit2/decl/graph.h -> build/lib.linux-x86_64-3.9/pygit2/decl
  copying pygit2/decl/commit.h -> build/lib.linux-x86_64-3.9/pygit2/decl
  copying pygit2/decl/stash.h -> build/lib.linux-x86_64-3.9/pygit2/decl
  copying pygit2/decl/refspec.h -> build/lib.linux-x86_64-3.9/pygit2/decl
  copying pygit2/decl/diff.h -> build/lib.linux-x86_64-3.9/pygit2/decl
  copying pygit2/decl/indexer.h -> build/lib.linux-x86_64-3.9/pygit2/decl
  copying pygit2/decl/blame.h -> build/lib.linux-x86_64-3.9/pygit2/decl
  copying pygit2/decl/remote.h -> build/lib.linux-x86_64-3.9/pygit2/decl
  copying pygit2/decl/clone.h -> build/lib.linux-x86_64-3.9/pygit2/decl
  copying pygit2/decl/submodule.h -> build/lib.linux-x86_64-3.9/pygit2/decl
  copying pygit2/decl/proxy.h -> build/lib.linux-x86_64-3.9/pygit2/decl
  copying pygit2/decl/checkout.h -> build/lib.linux-x86_64-3.9/pygit2/decl
  copying pygit2/decl/pack.h -> build/lib.linux-x86_64-3.9/pygit2/decl
  copying pygit2/decl/callbacks.h -> build/lib.linux-x86_64-3.9/pygit2/decl
  copying pygit2/decl/buffer.h -> build/lib.linux-x86_64-3.9/pygit2/decl
  copying pygit2/decl/transport.h -> build/lib.linux-x86_64-3.9/pygit2/decl
  copying pygit2/decl/errors.h -> build/lib.linux-x86_64-3.9/pygit2/decl
  copying pygit2/decl/types.h -> build/lib.linux-x86_64-3.9/pygit2/decl
  copying pygit2/decl/net.h -> build/lib.linux-x86_64-3.9/pygit2/decl
  copying pygit2/decl/merge.h -> build/lib.linux-x86_64-3.9/pygit2/decl
  copying pygit2/decl/index.h -> build/lib.linux-x86_64-3.9/pygit2/decl
  copying pygit2/decl/describe.h -> build/lib.linux-x86_64-3.9/pygit2/decl
  copying pygit2/decl/attr.h -> build/lib.linux-x86_64-3.9/pygit2/decl
  copying pygit2/decl/config.h -> build/lib.linux-x86_64-3.9/pygit2/decl
  copying pygit2/decl/revert.h -> build/lib.linux-x86_64-3.9/pygit2/decl
  copying pygit2/decl/oid.h -> build/lib.linux-x86_64-3.9/pygit2/decl
  copying pygit2/decl/repository.h -> build/lib.linux-x86_64-3.9/pygit2/decl
  copying pygit2/decl/strarray.h -> build/lib.linux-x86_64-3.9/pygit2/decl
  running build_ext
  generating cffi module 'build/temp.linux-x86_64-3.9/pygit2._libgit2.c'
  creating build/temp.linux-x86_64-3.9
  building 'pygit2._pygit2' extension
  creating build/temp.linux-x86_64-3.9/src
  x86_64-linux-gnu-gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -ffile-prefix-map=/build/python3.9-RNBry6/python3.9-3.9.2=. -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -fPIC -I/usr/local/include -I/usr/include/python3.9 -c src/blob.c -o build/temp.linux-x86_64-3.9/src/blob.o
  In file included from src/diff.h:34,
                   from src/blob.c:30:
  src/types.h:36:2: error: #error You need a compatible libgit2 version (1.4.x)
     36 | #error You need a compatible libgit2 version (1.4.x)
        |  ^~~~~
  error: command '/usr/bin/x86_64-linux-gnu-gcc' failed with exit code 1
  ----------------------------------------
  ERROR: Failed building wheel for pygit2
Failed to build pygit2
ERROR: Could not build wheels for pygit2 which use PEP 517 and cannot be installed directly
```

```

**Comment by whytewolf** [comment]
try version 1.6.1 that is the version i am using on centos 7 with ed certs and it is working fine for me. 

I'm not sure why you tried to compile when i explicitly told you to NOT use a compiled version as you will have to fight dependency issues. the main problem with what is happening is the libgit2 requirements on libssh2 causing problems.

**Comment by DaAwesomeP** [comment]
> did the pip install show that it was using the wheel? or did in compile it?

You asked if it used the wheel or compiled it. You did not say to not compile it. You only said not to use the system package. I harmlessly gave it a try. I have put back the wheel version and purged `libgit2-dev` from APT.

Regardless, the bundled shared library in the wheel was built against 1.4.0, so I suppose that mismatch could potentially be the issue with 1.9.0. pygit2 1.6.1 does appear to be built against libgit2 1.1.0.

Attempting 1.6.1 with ed25519 key, still did not work, although this time does not show as much error info:
```
$ python3 -m pip install --force pygit2==1.6.1
Collecting pygit2==1.6.1
  Downloading pygit2-1.6.1-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (4.6 MB)
     |████████████████████████████████| 4.6 MB 4.9 MB/s 
Collecting cffi>=1.4.0
  Using cached cffi-1.15.0-cp39-cp39-manylinux_2_12_x86_64.manylinux2010_x86_64.whl (444 kB)
Collecting pycparser
  Using cached pycparser-2.21-py2.py3-none-any.whl (118 kB)
Installing collected packages: pycparser, cffi, pygit2
  Attempting uninstall: pycparser
    Found existing installation: pycparser 2.21
    Uninstalling pycparser-2.21:
      Successfully uninstalled pycparser-2.21
  Attempting uninstall: cffi
    Found existing installation: cffi 1.15.0
    Uninstalling cffi-1.15.0:
      Successfully uninstalled cffi-1.15.0
  Attempting uninstall: pygit2
    Found existing installation: pygit2 1.9.0
    Uninstalling pygit2-1.9.0:
      Successfully uninstalled pygit2-1.9.0
Successfully installed cffi-1.15.0 pycparser-2.21 pygit2-1.6.1
$ systemctl restart salt-master.service 
$ salt-run fileserver.update
[ERROR   ] Unable to connect pusher: Stream is closed
[ERROR   ] Unable to connect pusher: Stream is closed
True
```

**Comment by whytewolf** [comment]
so the pull request would be against this repo. 

the file you would want to make the changes in is https://github.com/saltstack/salt/blob/master/doc/topics/tutorials/gitfs.rst

**Comment by garethgreenaway** [comment]
@DaAwesomeP Following up on this one, did you have a PR that you wanted to submit to fix the documentation around using pygit2 and Gitfs?  Thanks!

**Comment by DaAwesomeP** [comment]
I began to setup a new installation recently and became very confused with how the onedir system will affect this.

**Comment by cmcmarrow** [comment]
@DaAwesomeP what are you confused on onedir?

**Comment by DaAwesomeP** [comment]
OK, I just tried this again with onedir. I need to update my documentation fix to use `salt-pip`. I'll open a pull soon!

**Comment by DaAwesomeP** [comment]
@cmcmarrow It seems this entire page is still out-of-date and instructs users to install `pygit2` via OS package manage and not `salt-pip`. Is there already a timeline for this page to be updated?

https://docs.saltproject.io/en/latest/topics/tutorials/gitfs.html

**Comment by DaAwesomeP** [comment]
Closed via #63137, although that page desperately needs updated documentation for installing Python packages in general with onedir (a separate issue).
